File: project_euler_unified/problem_074/solution.py
"""
Project Euler Problem 74: Digit Factorial Chains

Count starting numbers below 1,000,000 whose digit factorial chain
contains exactly 60 non-repeating terms.

Uses memoized chain-length computation (Lemma: L(n) = 1 + L(f(n))
when n is not in a cycle).
"""

import math


def solve():
    fact = [math.factorial(d) for d in range(10)]

    def digit_fact_sum(n):
        s = 0
        while n > 0:
            s += fact[n % 10]
            n //= 10
        return s

    LIMIT = 1_000_000
    cache = {}

    def chain_length(n):
        if n in cache:
            return cache[n]

        chain = []
        seen = set()
        cur = n
        while cur not in seen:
            if cur in cache:
                total = len(chain) + cache[cur]
                for i, val in enumerate(chain):
                    cache[val] = total - i
                return cache.get(n, total)
            seen.add(cur)
            chain.append(cur)
            cur = digit_fact_sum(cur)

        total = len(chain)
        for i, val in enumerate(chain):
            cache[val] = total - i
        return total

    count = 0
    for start in range(1, LIMIT):
        if chain_length(start) == 60:
            count += 1

    return count


if __name__ == "__main__":
    answer = solve()
assert answer == 402
print(answer)

File: project_euler_unified/problem_074/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // Precompute digit factorials
    int fact[10];
    fact[0] = 1;
    for (int i = 1; i <= 9; i++) fact[i] = fact[i-1] * i;

    auto digit_fact_sum = [&](int n) -> int {
        if (n == 0) return 1;
        int s = 0;
        while (n > 0) {
            s += fact[n % 10];
            n /= 10;
        }
        return s;
    };

    const int LIMIT = 1000000;
    const int CACHE_SIZE = 3000001;  // f(n) <= 2,540,160 for n < 10^7
    vector<int> chain_len(CACHE_SIZE, 0);

    int count = 0;
    for (int start = 1; start < LIMIT; start++) {
        vector<int> chain;
        unordered_set<int> seen;
        int cur = start;
        while (seen.find(cur) == seen.end()) {
            if (cur < CACHE_SIZE && chain_len[cur] > 0) {
                int total = (int)chain.size() + chain_len[cur];
                for (int i = (int)chain.size() - 1; i >= 0; i--) {
                    int val = chain[i];
                    int len = total - i;
                    if (val < CACHE_SIZE && chain_len[val] == 0)
                        chain_len[val] = len;
                }
                if (total == 60) count++;
                goto next;
            }
            seen.insert(cur);
            chain.push_back(cur);
            cur = digit_fact_sum(cur);
        }
        {
            int total = (int)chain.size();
            for (int i = (int)chain.size() - 1; i >= 0; i--) {
                int val = chain[i];
                int len = total - i;
                if (val < CACHE_SIZE && chain_len[val] == 0)
                    chain_len[val] = len;
            }
            if (total == 60) count++;
        }
        next:;
    }

    cout << count << endl;
    return 0;
}
